src/sys/windows/udp.rs

```diff
@@ -137,17 +172,52 @@ impl UdpSocket {
                                            "failed to parse socket address"))
                     };
                     me.iocp.put_buffer(data);
-                    self.imp.schedule_read(&mut me);
+                    self.imp.schedule_read_from(&mut me);
+                    r
+                }
+            }
+            State::Error(e) => {
+                self.imp.schedule_read_from(&mut me);
+                Err(e)
+            }
+        }
+    }
+
+    pub fn recv(&self, mut buf: &mut [u8])
```

Comment:
Yeah, I like that a little bit better. We do a bit more work but that's already the case in the windows impl but it should be much cleaner and easier to reason about.